Realm/RLMCollection.mm

```diff
@@ -75,6 +78,40 @@ - (instancetype)initWithBackingCollection:(realm::object_store::Collection const
     return self;
 }
 
+- (instancetype)initWithBackingDictionary:(realm::object_store::Dictionary const&)backingDictionary
+                               dictionary:(RLMManagedDictionary *)dictionary
+                                classInfo:(RLMClassInfo&)info {
+    self = [super init];
+    if (self) {
+        _info = &info;
+        _realm = _info->realm;
+
+        if (_realm.inWriteTransaction) {
+            // Freezing a Results of `Dictionary::get_keys()` has very temperamental behavior.
+            // It also pins a version of a Realm which is something we do not want.
+            // It seems more efficient to store the keys in a temporary array.
+
+            // We still need some reference to the Results as we need to know
+            // if the parent object is valid or not.
+            _snapshot = backingDictionary.get_keys().snapshot();
+            for (id key in dictionary.allKeys) {
+                _dictionaryKeys.push_back(key);
+            }
+        }
+        else {
+            _snapshot = backingDictionary.get_keys();
+            for (id key in dictionary.allKeys) {
+                _dictionaryKeys.push_back(key);
+            }
```

Comment:
We need to keep a stable set of the keys for the enumeration, for the case that mutation happens during it. Snapshotting a Results of dictionary does nothing useful so we can't use that.

Reply:
Mutating a collection from inside a call to enumerateKeysAndObjectsUsingBlock is invalid and is allowed to break. We iterate over a snapshot for fast enumeration because live queries make it so that mutations the objects inside the collection can remove them from the query and indirectly mutate the collection being iterated, but that's not applicable here.